### PR TITLE
[FLINK-6352] [kafka] Timestamp-based offset configuration for FlinkKafkaConsumer

### PR DESCRIPTION
**Pull request by tzulitai**
## What is the purpose of the change

This PR is based on @zjureel's initial efforts on the feature in #3915.

This version mainly differs in that:
- When using timestamps to define the offset, the actual offset is eagerly determined in the `FlinkKafkaConsumerBase` class.
- The `setStartFromTimestamp` configuration method is defined in the `FlinkKafkaConsumerBase` class, with `protected` access. Kafka versions which support the functionality should override the method with `public` access.
- Timestamp is configured simply as a long value, and not a Java `Date`.

**Overall, the usage of the feature is as follows:**
```
FlinkKafkaConsumer011<String> consumer = new FlinkKafkaConsumer011<>(...);
consumer.setStartFromTimestamp(1515671654453L);

DataStream<String> stream = env.addSource(consumer);
...
```

Only versions 0.10 and 0.11 supports this feature.

**Semantics:**
- The provided timestamp cannot be larger than the current timestamp.
- For a partition, the earliest record which `record timestamp >= provided timestamp` is used as the starting offset.
- If the provided timestamp is larger than the latest record in a partition, that partition will simply be read from the head.
- For all new partitions that are discovered after the initial startup (due to scaling out Kafka), they are all read from the earliest possible record and the provided timestamp is not used.

## Brief change log

- d012826 @zjureel's initial efforts on the feature.
- 7ac07e8 Instead of lazily determining exact offsets for timestamp-based startup, the offsets are determined eagerly in `FlinkKafkaConsumerBase`. This commit also refactors the `setStartFromTimestamp` method to live in the base class.
- 32d46ef Change to just use long values to define timestamps, instead of using Java `Date`
- 7bb44a8 General improvement for the `runStartFromTimestamp` integration test.

## Verifying this change

New integration tests `Kafka010ITCase::testStartFromTimestamp` and `Kafka011ITCase::testStartFromTimestamp` verifies this new feature.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by tzulitai** [comment]
cc @zjureel

**Comment by aljoscha** [comment]
Initial question: why is the timestamp not used for newly discovered partitions?

**Comment by tzulitai** [comment]
@aljoscha this is the currently the case for any startup mode. Any partition discovered after the initial batch fetch is considered a new partition due to Kafka scale outs, and is therefore read from the record horizon. Startup modes apply only to existing partitions.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java [hunk 1]
```diff
@@ -343,6 +348,39 @@ public FlinkKafkaConsumerBase(
...
+	/**
+	 * Specifies the consumer to start reading partitions from a specified timestamp.
+	 * The specified timestamp must be before the current timestamp.
+	 * This lets the consumer ignore any committed group offsets in Zookeeper / Kafka brokers.
+	 *
+	 * <p>The consumer will look up the earliest offset whose timestamp is greater than or equal
+	 * to the specific timestamp from Kafka. If there's no such offset, the consumer will use the
+	 * latest offset to read data from kafka.
+	 *
+	 * <p>This method does not effect where partitions are read from when the consumer is restored
```
[the remark this answers is not in the document]
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java]: Will fix.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java [hunk 2]
```diff
@@ -343,6 +348,39 @@ public FlinkKafkaConsumerBase(
...
+	 * savepoint, only the offsets in the restored state will be used.
+	 *
+	 * @return The consumer object, to allow function chaining.
+	 */
+	// NOTE -
+	// This method is implemented in the base class because this is where the startup logging and verifications live.
+	// However, it is not publicly exposed since only newer Kafka versions support the functionality.
+	// Version-specific subclasses which can expose the functionality should override and allow public access.
+	protected FlinkKafkaConsumerBase<T> setStartFromTimestamp(long startupOffsetsTimestamp) {
+		checkNotNull(startupOffsetsTimestamp, "startupOffsetsTimestamp");
```
[the remark this answers is not in the document]
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java]: I'll change to a more meaningful message.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java [hunk 3]
```diff
@@ -343,6 +348,39 @@ public FlinkKafkaConsumerBase(
...
+
+		long currentTimestamp = System.currentTimeMillis();
+		checkArgument(startupOffsetsTimestamp <= currentTimestamp,
+			"Startup time[" + startupOffsetsTimestamp + "] must be before current time[" + currentTimestamp + "].");
```
[the remark this answers is not in the document]
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java]: 👌

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java [hunk 4]
```diff
@@ -441,28 +481,57 @@ public void open(Configuration configuration) throws Exception {
...
+							// default to group offset behaviour if the user-provided specific offsets
+							// do not contain a value for this partition
+							subscribedPartitionsToStartOffsets.put(seedPartition, KafkaTopicPartitionStateSentinel.GROUP_OFFSET);
+						}
+					}
+
+					break;
+				case TIMESTAMP:
+					if (startupOffsetsTimestamp == null) {
+						throw new IllegalArgumentException(
```
[the remark this answers is not in the document]
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java]: That makes sense, will change (including usage in existing code)

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 5]
```diff
@@ -1910,86 +1959,171 @@ public void cancel() {
...
+		readEnv
+			.addSource(consumer)
+			.map(new RichMapFunction<Tuple2<Integer, Integer>, Tuple2<Integer, Integer>>() {
+
+				private final int totalCount = parallelism * totalNumElements;
+				private int count = 0;
+
+				@Override
+				public Tuple2<Integer, Integer> map(Tuple2<Integer, Integer> value) throws Exception {
+					System.out.println(count);
```
[the remark this answers is not in the document]
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: Indeed, will remove.

**Comment by aljoscha** [comment]
Yes, that was my main concern. With a loop it could work, yes. 👍

**Comment by tzulitai** [comment]
@aljoscha I've addressed your comments and rebased the PR. Please have another look when you find the time, thanks a lot.

File: flink-connectors/flink-connector-kafka-0.10/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010FetcherTest.java [hunk 6]
```diff
@@ -129,9 +129,14 @@ public Void answer(InvocationOnMock invocation) {
 				schema,
 				new Properties(),
 				0L,
+<<<<<<< HEAD
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-0.10/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010FetcherTest.java]: Leftover merge markers?

**Comment by tzulitai** [comment]
@aljoscha sorry about the leftover merge markers, I've fixed them now.

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 7]
```diff
@@ -621,12 +621,70 @@ public void runStartFromSpecificOffsets() throws Exception {
...
+		long secondTimestamp = 0;
+		String topic = "";
+
+		// attempt to create an appended test sequence, where the timestamp of writing the appended sequence
+		// is assured to be larger than the timestamp of the original sequence.
+		final int maxRetries = 3;
+		int attempt = 0;
+		while (attempt != maxRetries) {
+			firstTimestamp = System.currentTimeMillis();
+			topic = writeSequence("runStartFromTimestamp", initialRecordsInEachPartition, parallelism, 1);
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: Ah, I just thought that we could have a simple loop there:

```
long secondTimestamp = System.currentTimeMillis();
while (secondTimestamp <= firstTimestamp) {
  Thread.sleep();
  secondTimestamp = System.currentTimeMillis();
}
```
what do you think?

**Comment by tzulitai** [comment]
Thanks for the review @aljoscha!

I'll proceed to merge this (while addressing your last comment) to `master` and `release-1.5`.

**Pull request closed without merging**